Spark: Add custom metric for number of deletes applied by a SparkScan

wypoon opened the pull request:
This is an extension of #4395.
Here we add a custom metric for the number of delete rows that have been applied in a scan of a format v2 table.

We introduce a counter in `BatchDataReader` and `RowDataReader`, that is incremented when a delete is applied. This counter is passed into `DeleteFilter`, and in the cases where we construct a `PositionDeleteIndex`, is passed into the implementation of the `PositionDeleteIndex`. In all the read paths, the counter is incremented whenever a delete is applied. When Spark calls `currentMetricsValues()` on a `PartitionReader`, which is a subclass of either `BatchDataReader` or `RowDataReader`, we get the current value of the counter and return that.

Tested manually by creating a format v2 table using each of Parquet, ORC, and Avro files, deleting and updating rows in the tables, and reading from the table. The expected number of deletes show up in the Spark UI.
Also extended existing unit tests (`DeleteReadTests`) to count the number of deletes applied during the scan.

<img width="529" alt="Screen Shot 2022-04-19 at 8 59 31 PM" src="https://user-images.githubusercontent.com/3925490/164147620-7085eafd-304d-45d1-aa53-0c6029638d48.png">

wypoon commented on core/src/main/java/org/apache/iceberg/deletes/DeleteCounter.java:
```diff
@@ -0,0 +1,33 @@
+
+package org.apache.iceberg.deletes;
+
+public class DeleteCounter {
```
This class is not strictly necessary. I could simply use an AtomicLong in its place. (I do not think, however, that the counter needs to be thread-safe, as each task will have its own counter.)

flyrain replied:
+1 for the new class.

wypoon commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnVectorWithFilter.java:
```diff
@@ -97,9 +97,18 @@ public byte[] getBinary(int rowId) {
+
   public static ColumnVector forHolder(VectorHolder holder, int[] rowIdMapping, int numRows) {
     return holder.isDummy() ?
-        new ConstantColumnVector(Types.IntegerType.get(), numRows, ((ConstantVectorHolder) holder).getConstant()) :
```
I do not know why the `ConstantColumnVector` is constructed with a hardcoded integer type. I observe that this code here can get called with `ConstantVectorHolder`s containing a constant of other primitive types. If `copy()` is called on an `InternalRow` in the `ColumnarBatch` containing this `ColumnVector`, then we get a `ClassCastException`. In the situations where I call `copy()`, the `ConstantColumnVector` contains either a boolean or an integer, so I only fix it for those possibilities.

wypoon replied:
@flyrain this code was added by you; can you please explain?

flyrain replied:
I copied some code from class `IcebergArrowColumnVector`, and didn't change its construction logic. But I agreed with you that we shouldn't hard-code to int type.

wypoon commented:
@rdblue @RussellSpitzer would you please review?

RussellSpitzer commented on data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:
```diff
@@ -172,7 +185,8 @@ public CloseableIterable<T> findEqualityDeleteRows(CloseableIterable<T> records)
       @Override
       protected boolean shouldKeep(T item) {
-        return deletedRows.test(item);
+        boolean keep = deletedRows.test(item); // true means row is deleted
```
This probably can be reverted?

wypoon replied:
Reverted.

RussellSpitzer commented on data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:
```diff
@@ -188,31 +202,52 @@ private CloseableIterable<T> applyEqDeletes(CloseableIterable<T> records) {
   }
 
   public PositionDeleteIndex deletedRowPositions() {
+    LOG.debug("calling deletedRowPositions()");
```
Probably remove these messages?

wypoon replied:
Removed.

RussellSpitzer commented on data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:
```diff
@@ -226,17 +261,18 @@ private CloseableIterable<T> applyPosDeletes(CloseableIterable<T> records) {
   }
 
   private CloseableIterable<Record> openDeletes(DeleteFile deleteFile, Schema deleteSchema) {
+    LOG.trace("opening delete file {}", deleteFile.path());
```
Should start with a Capital

wypoon replied:
Done.

RussellSpitzer commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:
```diff
@@ -47,24 +48,34 @@
   @Override
   CloseableIterator<ColumnarBatch> open(FileScanTask task) {
     DataFile file = task.file();
+    LOG.debug("opening data file {}", file.path());
```
Start with a capital

wypoon replied:
Done.

RussellSpitzer commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:
```diff
@@ -75,6 +86,7 @@ CloseableIterator<ColumnarBatch> open(FileScanTask task) {
     InputFile location = getInputFile(task);
     Preconditions.checkNotNull(location, "Could not find InputFile associated with FileScanTask");
     if (task.file().format() == FileFormat.PARQUET) {
+      LOG.debug("data file is parquet; calling deleteFilter to get SparkDeleteFilter ...");
```
Capital? But maybe remove this one, not sure it's adding a lot of value since you could just add the line in the `deleteFilter` code

wypoon replied:
Removed.

RussellSpitzer commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:
```diff
@@ -126,7 +138,8 @@ CloseableIterator<ColumnarBatch> open(FileScanTask task) {
 
   private SparkDeleteFilter deleteFilter(FileScanTask task) {
-    return task.deletes().isEmpty() ? null : new SparkDeleteFilter(task, table().schema(), expectedSchema);
+    LOG.debug("is task.deletes() empty? {}", task.deletes().isEmpty());
```
I would probably just break up the ternary operator below and log when creating the Delete filter?

wypoon replied:
Done.

RussellSpitzer commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:
```diff
@@ -46,25 +47,34 @@
   @Override
   CloseableIterator<InternalRow> open(FileScanTask task) {
-    SparkDeleteFilter deletes = new SparkDeleteFilter(task, tableSchema, expectedSchema);
+    LOG.debug("calling open, constructing SparkDeleteFilter ...");
```
Capital here, but i'm not sure this is a helpful debug statement either? Probably add a line in the DeleteFilter Constructor instead?

wypoon replied:
Removed.

RussellSpitzer commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkBatchQueryScan.java:
```diff
@@ -93,6 +93,11 @@ class SparkBatchQueryScan extends SparkScan implements SupportsRuntimeFiltering
 
+  @Override
+  public String description() {
+    return "IcebergBatchScan " + super.description();
```
Not sure how these changes are related to the PR

RussellSpitzer commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkCopyOnWriteScan.java:
```diff
@@ -77,6 +77,11 @@ class SparkCopyOnWriteScan extends SparkScan implements SupportsRuntimeFiltering
 
+  @Override
+  public String description() {
+    return "IcebergCopyOnWriteScan " + super.description();
```
Probably separate this into another PR

RussellSpitzer commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/metrics/NumDeletes.java:
```diff
@@ -0,0 +1,45 @@
+
+  @Override
+  public String description() {
+    return "number of row deletes applied";
```
Capital first

wypoon replied:
I use a lowercase "number" to be consistent with other (standard) metrics appearing in the Spark UI (e.g., "number of output rows"). It would look odd otherwise.

RussellSpitzer replied:
that's fine then, I forgot Spark did that

RussellSpitzer commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/metrics/NumDeletes.java:
```diff
@@ -0,0 +1,45 @@
+import java.text.NumberFormat;
+import org.apache.spark.sql.connector.metric.CustomMetric;
+
+public class NumDeletes implements CustomMetric {
```
Should probably extend CustomSumMetric

https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/connector/metric/CustomSumMetric.html

wypoon replied:
I am aware of `CustomSumMetric` but there is no benefit in extending it instead of `CustomMetric`, since I would still override its `aggregateTaskMetrics` method as that returns the `String` representation of the long instead of a locale-formatted string as I do (and as other longs are shown in the Spark UI). So if there are 1,000 deletes, it should show as "1,000" and not "1000".

wypoon replied:
I can open a Spark PR to change `CustomSumMetric`, but for now I'd just extend `CustomMetric`.
The thing is, Spark instantiates your `CustomMetric` class and using reflection, calls its `aggregateTaskMetrics` method to get the value to display in the UI. However, even if you extend `CustomSumMetric` and don't override `aggregateTaskMetrics`, Spark still needs to instantiate your `CustomMetric` class, so you can't get away from putting your `CustomMetric` class in the classpath of the Spark History Server. If there was a way to just use `CustomSumMetric` (and live with its deficient output formatting) and not have to provide your `CustomMetric` class to the SHS, I'd accept that, but there isn't. Thus I really don't see any benefit to extending `CustomSumMetric`.

RussellSpitzer commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/metrics/TaskNumDeletes.java:
```diff
@@ -0,0 +1,40 @@
+import org.apache.spark.sql.connector.metric.CustomTaskMetric;
+
+public class TaskNumDeletes implements CustomTaskMetric {
+  private long value;
```
nit: final

wypoon replied:
Done.

RussellSpitzer commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:
```diff
@@ -139,18 +151,55 @@ protected void dropTable(String name) {
+        TableProperties.SPLIT_LOOKBACK_DEFAULT,
+        TableProperties.SPLIT_OPEN_FILE_COST_DEFAULT);
+
+    long delCount = 0L;
```
This feels like we are subverting the intent of this test which is to use the DF reader method to actually read the table. I think we'll need another approach for counting deletes or possibly just explicitly state how many deletes there are in tests that need to check it. At least until we have a method of reading a table with a user readable marker for deleted rows like `_isDeleted`

wypoon replied:
I could replace 
```
    CloseableIterable<CombinedScanTask> tasks = TableScanUtil.planTasks(
        table.newScan().planFiles(),
        TableProperties.METADATA_SPLIT_SIZE_DEFAULT,
        TableProperties.SPLIT_LOOKBACK_DEFAULT,
        TableProperties.SPLIT_OPEN_FILE_COST_DEFAULT);
```
with
```
    SparkScanBuilder scanBuilder = new SparkScanBuilder(spark, table, CaseInsensitiveStringMap.empty());
    scanBuilder.pruneColumns(sparkSchema);
    SparkScan scan = (SparkScan) scanBuilder.buildMergeOnReadScan();
    List<CombinedScanTask> tasks = scan.tasks();
```
That is the path that a Spark read takes. I am also doing most of the rest of what Spark does once we have the tasks and hand them to Spark in the form of `InputPartition`s, which is in each `InputPartition`, get a `PartitionReader` (which is a subclass of either `BatchDataReader` or `RowDataReader`) and iterate through the reader. I just don't assemble the `InternalRow`s into a `DataFrame`.

RussellSpitzer replied:
That is the path it takes at this moment, but that's the issue here, we want this test check the output of
```    
Dataset<Row> df = spark.read()
        .format("iceberg")
        .load(TableIdentifier.of("default", name).toString())
        .selectExpr(columns);
```

Having this copy the implementation details decouples the test from what actually may be happening we change the code in the future

wypoon replied:
Here is another option:
I'll leave `TestSparkReaderDeletes` alone, and instead create a new class that also extends `DeleteReadTests` (or even `TestSparkReaderDeletes`) and override `rowSet` there. It is perhaps also not necessary to run all the tests in `DeleteReadTests` or `TestSparkReaderDeletes` in this new class. I had also increased the parameter space of `TestSparkReaderDeletes` in order to cover all 3 file formats, but I can dial it back and distribute the increased coverage to the new class. Basically, there will be no loss of test coverage.

wypoon commented:
@RussellSpitzer thank you for reviewing!
This PR is on top of #4395, but since that is not merged, all its changes are here.
I'll clean up the logging. It is true that their original purpose was for debugging/testing, but I think some of them are still potentially helpful.

kbendick commented on core/src/main/java/org/apache/iceberg/deletes/BitmapPositionDeleteIndex.java:
```diff
@@ -40,7 +46,11 @@ public void delete(long posStart, long posEnd) {
   @Override
   public boolean isDeleted(long position) {
-    return roaring64Bitmap.contains(position);
+    boolean posIsDeleted = roaring64Bitmap.contains(position);
+    if (counter != null && posIsDeleted) {
+      counter.increment();
+    }
```
Given that we have defined the `DeleteCounter` interface ourselves, does it make sense to define a default `NullDeleteCounter` that is a no-op?

There are a number of places where we check `if (counter != null)`.

kbendick replied:
Also, is there any concern that this makes calls to `isDeleted` not idempotent?

wypoon replied:
There are existing tests that have no need for counting deletes, that call some static methods in `Deletes`. I kept those forms of the methods and had them call new forms that take a `DeleteCounter`, passing a null `DeleteCounter`. That is the reason why a `DeleteCounter` field might be null. The way the code is now, if we decide to do without `DeleteCounter` and simply use an `AtomicLong`, then it can be updated very easily. I do understand your suggestion for a `NullDeleteCounter`.

You raise a good point about `PositionDeleteIndex#isDeleted` not being idempotent now. I have checked where it is called, and it is not a concern. We should document this point though, to prevent it being a problem in future code. Before I ended up introducing a delete counter into the `PositionDeleteIndex` implementation and incrementing it when `isDeleted` is called, I considered other approaches, but there was one read path in particular where this seemed to the best solution.

wypoon replied:
Added a note to the javadoc for `PositionDeleteIndex#isDeleted`.

RussellSpitzer replied:
Shouldn't we be able to determine the number of deleted rows here by checking the number of entries in the bitmap? So shouldn't we be able to just count in the "delete methods"? Also do we have to pass through the counter here? Seems like we could just get this value out of the index when we are done building it instead of passing through the counter into the index?

wypoon replied:
You are right that one can get the number of deleted positions in the `PositionDeleteIndex` by `roaring64Bitmap.getLongCardinality()`.
However, having one counter keeps the bookkeeping simple and easy to reason about.
There are multiple paths through the code that can be taken, where deletes may be applied, and the unifying thing here is that the same counter is passed into all these places.
It is more difficult to reason through the counting if I have to account for the `PositionDeleteIndex` cases separately and I would have to keep references to the `PositionDeleteIndex` instances (I haven't looked in depth into this). I think it would be messy.

RussellSpitzer replied:
I think having "isDeleted" have a side effect ruins the book keeping simplicity for me. I'd much rather if we were to do this counter approach we would the "delete" methods rather than the "isDeleted" method. The delete method already has a side effect and we would expect it to be the place where the count would happen as opposed to "isDeleted" which we have to add a JavaDoc note to for illustrating the side effect behavior. 

I still am not really sold on passing through the counter here, it just seems to me like we are complicating the internals of this class and we don't really need to since like we've already discussed, the value we are looking for is already calculated even without us passing through  an object. I'm also not a big fan of behaviors being dependent on objects being null or not.

But if other reviewers feel differently I could go along with this approach as well.

wypoon replied:
Since it has been almost 4 weeks since I touched this (it seems that's how long the last update has gone unreviewed), I had forgotten the details of what I had investigated and had done. I went through the code paths again, and I see now that there is just one place where `PositionDeleteIndex#delete()` is called and where I need to track the delete count in the `PositionDeleteIndex`. So your suggestion is very feasible and not messy as I feared!
Thank you for the suggestion. I'll update the PR tomorrow.

wypoon replied:
@kbendick I considered using an `Optional<DeleteCounter>` instead of a `DeleteCounter` that could be null, but it's six of one and half a dozen of the other. In the end, I adopted your suggestion to have a no-op `DeleteCounter`.

wypoon commented on core/src/main/java/org/apache/iceberg/deletes/Deletes.java:
```diff
@@ -133,11 +158,17 @@ protected EqualitySetDeleteFilter(Function<T, StructLike> extractEq,
     protected boolean shouldKeep(T row) {
-      return !deletes.contains(extractEqStruct.apply(row));
+      boolean keep = !deletes.contains(extractEqStruct.apply(row));
+      if (!keep) {
```
I have removed these trace messages.

wypoon commented on data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:
```diff
@@ -67,22 +72,26 @@
     for (DeleteFile delete : deletes) {
       switch (delete.content()) {
         case POSITION_DELETES:
+          LOG.trace("adding position delete file {} to filter", delete.path());
```
Done.

wypoon commented:
@RussellSpitzer I have cleaned up the logging, and reworked the unit tests, on top of the updates to #4395.

RussellSpitzer commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeleteCount.java:
```diff
@@ -0,0 +1,109 @@
+  @Parameterized.Parameters(name = "format = {0}, vectorized = {1}")
+  public static Object[][] parameters() {
+    return new Object[][] {
+        new Object[] {"parquet", true},
```
Should there be a "parquet", false here as well?

wypoon replied:
This is no longer applicable as I have reverted splitting out testing delete count into a separate test class.

RussellSpitzer commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeleteCount.java:
```diff
@@ -0,0 +1,109 @@
+    Types.StructType type = schema.asStruct();
+    StructLikeSet set = StructLikeSet.create(type);
+
+    SparkScanBuilder scanBuilder = new SparkScanBuilder(spark, table, CaseInsensitiveStringMap.empty());
```
Instead of doing this would it be possible for us to use a custom SparkListener to just grab the status at the end of the job and check those values? I think it is a safer approach than reimplementing the read path here.

We could also just modify the SQLAppStatusListener? or something like that

RussellSpitzer commented:
@flyrain could you take a look at this?

wypoon commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:
```diff
@@ -94,11 +141,15 @@ public static void startMetastoreAndSpark() {
 
     spark = SparkSession.builder()
         .master("local[2]")
+        .config("spark.appStateStore.asyncTracking.enable", false)
```
The default is true, which means that when the `SQLAppStatusListener` receives a `SparkListenerSQLExecutionEnd` event, it asks its `ElementTrackingStore` to asynchronously aggregate the metrics. This async call sometimes results in the metric we need not being available. To make the tests more predictable, ensure that a synchronous call to aggregate metrics is used.

wypoon commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:
```diff
@@ -152,9 +220,53 @@ public StructLikeSet rowSet(String name, Table table, String... columns) {
+    // Get the executionId of the query we just executed
+    List<Long> executed = listener.executed();
+    long lastExecuted = executed.get(executed.size() - 1);
+    // Ensure that the execution end was registered
```
Wait for the `SparkListenerSQLExecutionEnd` to be received. If our `SparkListener` has received the event, the `SQLAppStatusListener` should have received it too and called its `ElementTrackingStore` to aggregate the metrics.

RussellSpitzer replied:
Oh this is great, Looking into the apis you dug up I think there may be an even easier way to do this. Sorry I didn't realize you could do this before

Looking into sharedState you can do something like
```scala
// Some operation
scala> spark.range(1, 100).withColumn("x", col("id")).withColumnRenamed("id", "y").writeTo("local.default.test").append

// UI Metrics
scala> spark.sharedState.statusStore.executionsList.last.metrics
res41: Seq[org.apache.spark.sql.execution.ui.SQLPlanMetric] = List(SQLPlanMetric(number of output rows,25,sum), SQLPlanMetric(duration,24,timing))

// Description
scala> spark.sharedState.statusStore.executionsList.last.description
res42: String = append at <console>:23

// Execution id
scala> spark.sharedState.statusStore.executionsList.last.executionId
res47: Long = 6
```

So this block could be like
```java
  private long lastExecutedMetric(String metricName) {
    SQLExecutionUIData lastExecution = spark.sharedState().statusStore().executionsList().last();
    Option<SQLPlanMetric> sqlPlanMetric = lastExecution.metrics().find(metric -> metric.name().equals(metricName));
    Assert.assertTrue("Metric xxxx not present in last execution", sqlPlanMetric.isDefined());
    long metricId = sqlPlanMetric.get().accumulatorId();

    // Refresh metricValues, they will remain null until the execution is complete and metrics are aggregated
    int attempts = 3;
    while (lastExecution.metricValues() == null && attempts > 0) {
      try {
        Thread.sleep(100);
        attempts -= 1;
      } catch (InterruptedException e) {
        throw new RuntimeException(e);
      }
      lastExecution = spark.sharedState().statusStore().execution(lastExecution.executionId()).get();
    }
    Assert.assertNotNull(String.format("Metric %s was never finalized", metricName), lastExecution.metricValues());
    return Long.parseLong(lastExecution.metricValues().get(metricId).get());
  }
  ```
  
  Then we just have
  ```
    @Override
  protected void checkDeleteCount(long expectedDeletes) {
    Assert.assertEquals("Unexpected number of deletes recorded", expectedDeletes,
        lastExecutedMetric(NumDeletes.DISPLAY_STRING));
  }
  ```
  
  And we can drop the other state in the test class.

RussellSpitzer replied:
Also can you please add this to the Num Splits PR so we can have a few tests in there as well? This will be really useful for all of our future metric additions

wypoon commented:
@kbendick and @RussellSpitzer thank you for your reviews and suggestions.
I have removed `DeleteCounter` from `BitmapPositionDeleteIndex`. I have implemented the suggestion to extract the custom metric from Spark after running a DataFrame query (although this turned out to be trickier than expected due to async issues (that did not surface locally but surfaced in the CI)). I think the PR is much better now.

wypoon commented on data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:
```diff
@@ -225,18 +257,19 @@ private CloseableIterable<T> applyPosDeletes(CloseableIterable<T> records) {
     List<CloseableIterable<Record>> deletes = Lists.transform(posDeletes, this::openPosDeletes);
 
     // if there are fewer deletes than a reasonable number to keep in memory, use a set
-    if (posDeletes.stream().mapToLong(DeleteFile::recordCount).sum() < setFilterThreshold) {
-      return Deletes.filter(records, this::pos, Deletes.toPositionIndex(filePath, deletes));
+    if (posDeletes.stream().mapToLong(DeleteFile::recordCount).sum() < streamFilterThreshold) {
+      return Deletes.filter(records, this::pos, Deletes.toPositionIndex(filePath, deletes, counter));
     }
 
-    return Deletes.streamingFilter(records, this::pos, Deletes.deletePositions(filePath, deletes));
+    return Deletes.streamingFilter(records, this::pos, Deletes.deletePositions(filePath, deletes), counter);
```
This is a code path that was not previously exercised by the unit tests.
We add a configuration that allows us to control the `streamFilterThreshold` and cause this code path to be taken.

RussellSpitzer commented on data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:
```diff
@@ -54,35 +56,42 @@
 public abstract class DeleteFilter<T> {
-  private static final long DEFAULT_SET_FILTER_THRESHOLD = 100_000L;
+  private static final Logger LOG = LoggerFactory.getLogger(DeleteFilter.class);
+  public static final long DEFAULT_STREAM_FILTER_THRESHOLD = 100_000L;
```
Can we extract all of these set/streamFilter changes out into another PR?

wypoon replied:
I will revert the last commit from this PR. If this PR can be merged without that, I'll open a separate PR once it is merged, consisting of that change.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/deletes/Deletes.java:
```diff
@@ -84,27 +89,46 @@ public static StructLikeSet toEqualitySet(CloseableIterable<StructLike> eqDelete
     try (CloseableIterable<Long> deletes = posDeletes) {
       PositionDeleteIndex positionDeleteIndex = new BitmapPositionDeleteIndex();
       deletes.forEach(positionDeleteIndex::delete);
+      counter.increment(((BitmapPositionDeleteIndex) positionDeleteIndex).numberOfPositionsDeleted());
```
we can the skip the cast here if we just keep the type on 102

wypoon replied:
If we add `numberOfPositionsDeleted()` in the `PositionDeleteIndex` interface, then we don't need the cast either.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/deletes/BitmapPositionDeleteIndex.java:
```diff
@@ -47,4 +47,8 @@ public boolean isDeleted(long position) {
     return roaring64Bitmap.isEmpty();
   }
+
+  public long numberOfPositionsDeleted() {
```
I think this may be ok to include in the interface itself, I can't think of an implementation which couldn't track this?

wypoon replied:
Ok, I can put `numberOfPositionsDeleted()` in the `PositionDeleteIndex` interface.

wypoon replied:
Hmm, turns out there is a `TestSparkParquetReadMetadataColumns.CustomizedPositionDeleteIndex` that implements `PositionDeleteIndex`. I'll leave the interface alone then, since I don't want to change unrelated tests.

flyrain replied:
It makes sense to put it in the interface. The change of `CustomizedPositionDeleteIndex` is minor. 
Minor suggestion: `numberOfPositionsDeleted` -> `numberOfPositionDeletes`

flyrain replied:
Looking a bit more, we don't need this due to https://github.com/apache/iceberg/pull/4588/files#r905365603.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/deletes/NonCounter.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * A dummy counter that does not increment its count.
+ */
+class NonCounter implements DeleteCounter {
```
Not sure why we need this class, Is there a case when we really don't want to count? I would imagine perf wise the difference between counting and not counting is pretty small?

wypoon replied:
@kbendick suggested a no-op version of the `DeleteCounter` instead of using null. There are legacy uses where counting is not needed, but I tend to agree that it doesn't hurt to use a single `DeleteCounter` throughout and that `NonCounter` is probably not needed.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/deletes/Deletes.java:
```diff
@@ -84,27 +89,46 @@ public static StructLikeSet toEqualitySet(CloseableIterable<StructLike> eqDelete
     }
   }
 
+  public static <T extends StructLike> PositionDeleteIndex toPositionIndex(CharSequence dataLocation,
```
Dow we need these polymorphisms? Or is this just to keep apis?

wypoon replied:
Yes, there are there to support existing API calls.

RussellSpitzer commented on data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:
```diff
@@ -67,22 +72,26 @@
 
-  protected DeleteFilter(String filePath, List<DeleteFile> deletes, Schema tableSchema, Schema requestedSchema) {
+  protected DeleteFilter(String filePath, List<DeleteFile> deletes, Schema tableSchema, Schema requestedSchema,
+      DeleteCounter counter) {
```
Do we need to pass through the counter here? Seems like this should just be an internal member of delete filter that is created with the delete filter? Or is this just because we have no other way of getting the counter information back out of the applied delete filter?

wypoon replied:
Yes, we need to pass the `DeleteCounter` in. The `DeleteCounter` is created in either `RowDataReader` or `BatchDataReader` and in their `open(FileScanTask)` method,  an instance of their `SparkDeleteFilter` static nested class is created for that `FileScanTask` with the `DeleteCounter`. We need a single counter in either `RowDataReader` or `BatchDataReader` which is used to aggregate the delete count over all the `FileScanTask`s.

wypoon replied:
Update: The `DeleteCounter` is now created in `BaseReader` and passed to the instance of `BaseReader.SparkDeleteFilter` constructed in the `open(FileScanTask)` method `RowDataReader`/`BatchDataReader`. The argument still stands: we need a single counter in the reader which is used to aggregate the delete count over all the `FileScanTask`s.

RussellSpitzer commented on data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:
```diff
@@ -96,6 +105,10 @@ protected DeleteFilter(String filePath, List<DeleteFile> deletes, Schema tableSc
     this.posAccessor = requiredSchema.accessorForField(MetadataColumns.ROW_POSITION.fieldId());
   }
 
+  protected DeleteFilter(String filePath, List<DeleteFile> deletes, Schema tableSchema, Schema requestedSchema) {
```
Is this just for legacy compatibility? Do we have any users of this here? as a protected method I would think we would be free to drop it if we wanted to.

wypoon replied:
This constructor without the `DeleteCounter` is called by `GenericDeleteFilter` and `FlinkDeleteFilter`.

RussellSpitzer commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchReader.java:
```diff
@@ -195,20 +199,8 @@ int[] initEqDeleteRowIdMapping(int numRowsToRead) {
-      }
-
+      LOG.debug("Applying equality deletes to row id mapping");
+      int currentRowId = deletes.applyEqDeletes(it, rowIdMapping);
```
Is this just a refactoring?

wypoon replied:
Yes and no. There doesn't seem to be a natural way to get the `DeleteCounter` to `ColumnarBatchReader`. In this method, the `rowIdMapping` int array is updated by applying the equality deletes in the `DeleteFilter` `deletes`. As each row is tested, if it is deleted, we need to increment the `DeleteCounter`. The `DeleteFilter` has the `DeleteCounter`, so I moved the logic completely over to `DeleteFilter`. The alternative is to expose `DeleteFilter`'s `DeleteCounter` so that it can be used here, but I think it is better to encapsulate that in `DeleteFilter`.

flyrain replied:
I'd prefer not to move this logic to `DeleteFilter` since it is more relevant here. We may expose DeleteFilter's DeleteCounter as you suggested.

RussellSpitzer commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:
```diff
@@ -139,7 +193,21 @@ protected void dropTable(String name) {
 
   @Override
-  public StructLikeSet rowSet(String name, Table table, String... columns) {
+  protected boolean countDeletes() {
```
Instead of the apis here, could we just modify "extractDeleteCount()" 

to instead be something like "
```java 
long lastMetricCount(String metricName) {
// Get executionID of last run thing
// Find metricName
// return value for MetricName
}
```

This way we can avoid some state in the job?

wypoon replied:
I removed the `deleteCount` field and the `setDeleteCount(long)` method, and the `deleteCount()` method now simply calls a `lastExecutedMetricValue(...)` method along the lines you suggest.

RussellSpitzer reviewed (commented):
Left some more general structural comments. I think Ideally I want to push that state up as high as we can in the stack. Anything which doesn't actually need to receive a counter in it's constructor should avoid it imho.

I think we would also really benefit from @flyrain doing a full review on this as well especially now that we have the "markDelete" pathway as well. I assume for that we probably will just skip counting deletes since we don't really care.

wypoon commented:
> I think we would also really benefit from @flyrain doing a full review on this as well especially now that we have the "markDelete" pathway as well. I assume for that we probably will just skip counting deletes since we don't really care.

@RussellSpitzer I saw that `Deletes` and `DeleteFilter` have changed in master since I updated this PR (so this PR no longer merges), but I haven't had the time to understand the changes and how to reconcile this PR with them.
I don't understand your statement that "we probably will just skip counting deletes since we don't really care." Are you saying that this PR is no longer worthwhile??

RussellSpitzer commented:
> > I think we would also really benefit from @flyrain doing a full review on this as well especially now that we have the "markDelete" pathway as well. I assume for that we probably will just skip counting deletes since we don't really care.
> 
> @RussellSpitzer I saw that `Deletes` and `DeleteFilter` have changed in master since I updated this PR (so this PR no longer merges), but I haven't had the time to understand the changes and how to reconcile this PR with them. I don't understand your statement that "we probably will just skip counting deletes since we don't really care." Are you saying that this PR is no longer worthwhile??

No! I think this PR is very important, there is just now a second read path with `_is_deleted` metadata column which actually returns deleted rows. When we follow that path I'm not sure it's important for us to count the deleted rows since we'll be returning them anyway, I could go either way. For the path when we don't actually return deleted rows (the normal read path) we definitely want this metric!

wypoon commented:
> > > I think we would also really benefit from @flyrain doing a full review on this as well especially now that we have the "markDelete" pathway as well. I assume for that we probably will just skip counting deletes since we don't really care.
> > 
> > 
> > @RussellSpitzer I saw that `Deletes` and `DeleteFilter` have changed in master since I updated this PR (so this PR no longer merges), but I haven't had the time to understand the changes and how to reconcile this PR with them. I don't understand your statement that "we probably will just skip counting deletes since we don't really care." Are you saying that this PR is no longer worthwhile??
> 
> No! I think this PR is very important, there is just now a second read path with `_is_deleted` metadata column which actually returns deleted rows. When we follow that path I'm not sure it's important for us to count the deleted rows since we'll be returning them anyway, I could go either way. For the path when we don't actually return deleted rows (the normal read path) we definitely want this metric!

Thanks for the explanation. I haven't followed the changes that recently went in. Your explanation helps put them in context for me. I'll rebase on master and try to address your current feedback.

flyrain commented:
Sorry, I haven't got chance to review. The changes of class `Deletes` and `DeleteFilter` are introduced by my PR #4683, which can read deleted rows by using metadata column `_deleted`. It may affect how we collect the deletes metrics. Would recommend to take a look at #4683, @wypoon. I will also review this PR a bit later.

RussellSpitzer commented:
Benchmarking could be good, although I'm gonna be real surprised if it's that heavy. I'll take bets on undetectable difference ... I'm assuming all the function calls will end up getting inlined during runtime optimizations

wypoon commented on data/src/main/java/org/apache/iceberg/data/DeleteFilter.java:
```diff
@@ -212,7 +244,7 @@ public PositionDeleteIndex deletedRowPositions() {
     if (deleteRowPositions == null) {
       List<CloseableIterable<Record>> deletes = Lists.transform(posDeletes, this::openPosDeletes);
-      deleteRowPositions = Deletes.toPositionIndex(filePath, deletes);
+      deleteRowPositions = Deletes.toPositionIndex(filePath, deletes, counter);
```
Actually, this eventually calls `Deletes.toPositionIndex(CloseableIterable<Long>, DeleteCounter)` where we do
```
      PositionDeleteIndex positionDeleteIndex = new BitmapPositionDeleteIndex();
      deletes.forEach(positionDeleteIndex::delete);
      counter.increment(((BitmapPositionDeleteIndex) positionDeleteIndex).numberOfPositionsDeleted());
      return positionDeleteIndex;
```
That method is called by other callers besides this. I think it is best to do the `counter.increment(positionDeleteIndex.numberOfPositionsDeleted())` right where `PositionDeleteIndex#delete` is called. I think the bookkeeping is easier that way.

wypoon commented:
I'm working on rebasing on master, reconciling the conflicts with #4683.

wypoon commented:
@RussellSpitzer @flyrain I have rebased on master and reconciled the conflicts with #4683. Counting deletes only happen for the normal read path, not the new path added by #4683.
Russell, I have adopted your suggestions for `TestSparkReaderDeletes`. For the other questions/points you raised, I have answered them earlier.
I don't think this work has any performance impact. The costs of incrementing an in-memory counter are dwarfed by the costs of opening and reading delete files.

flyrain commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchReader.java:
```diff
@@ -133,6 +136,7 @@ void initRowIdMapping(int numRowsToRead) {
 
     Pair<int[], Integer> posDelRowIdMapping(int numRowsToRead) {
       if (deletes != null && deletes.hasPosDeletes()) {
+        LOG.debug("Building row id mapping from positional deletes");
```
Why do we need a debug log here?

flyrain commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/metrics/NumDeletes.java:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  @Override
+  public String aggregateTaskMetrics(long[] taskMetrics) {
+    long sum = initialValue;
+    for (int i = 0; i < taskMetrics.length; i++) {
+      sum += taskMetrics[i];
+    }
+    return NumberFormat.getIntegerInstance().format(sum);
+  }
```
How about extending from CustomSumMetric? So that you don't have to reimplement this method.

flyrain commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:
```diff
@@ -46,30 +47,39 @@
   private final Schema expectedSchema;
   private final String nameMapping;
   private final boolean caseSensitive;
+  private final DeleteCounter counter;
```
I'd also suggest to move this to `BaseDataReader`.

flyrain commented on core/src/main/java/org/apache/iceberg/deletes/Deletes.java:
```diff
@@ -202,17 +233,24 @@ boolean isDeleted(T row) {
         @Override
         protected boolean shouldKeep(T item) {
-          return !isDeleted(item);
+          boolean deleted = isDeleted(item);
+          if (deleted) {
+            counter.increment();
+          }
+          return !deleted;
```
This may have perf impact since it is per-row basis. I don't think it is big though, but would recommend to use jmh to micro benchmark it.

wypoon replied:
I ran `IcebergSourceParquetPosDeleteBenchmark` before and after this change, and the numbers for `readIceberg` show hardly any difference.

flyrain commented on core/src/main/java/org/apache/iceberg/deletes/Deletes.java:
```diff
@@ -75,8 +75,21 @@ public static <T> CloseableIterable<T> markDeleted(CloseableIterable<T> rows, Pr
+    Filter<T> remainingRowsFilter = new Filter<T>() {
+      @Override
+      protected boolean shouldKeep(T item) {
+        boolean deleted = isDeleted.test(item);
+        if (deleted) {
+          counter.increment();
+        }
+        return !deleted;
```
The same perf concern here.

flyrain commented on core/src/main/java/org/apache/iceberg/deletes/Deletes.java:
```diff
@@ -90,27 +103,45 @@ public static StructLikeSet toEqualitySet(CloseableIterable<StructLike> eqDelete
-      PositionDeleteIndex positionDeleteIndex = new BitmapPositionDeleteIndex();
+      BitmapPositionDeleteIndex positionDeleteIndex = new BitmapPositionDeleteIndex();
       deletes.forEach(positionDeleteIndex::delete);
+      counter.increment(positionDeleteIndex.numberOfPositionsDeleted());
```
This is not the right way. positionDeleteIndex has all pos deletes for the data file, but not necessary all deletes are used in a read. We need to do the same thing as batch eq delete(line 223 in DeleteFilter.java)

wypoon replied:
Thanks for pointing this out. I have fixed this.

flyrain reviewed (commented):
@wypoon thanks for working on this. I believe this is pretty nice to have. Left some comments. Sorry for the delay.

wypoon commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/metrics/NumSplits.java:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  @Override
+  public String aggregateTaskMetrics(long[] taskMetrics) {
+    long sum = initialValue;
+    for (int i = 0; i < taskMetrics.length; i++) {
+      sum += taskMetrics[i];
+    }
+    return NumberFormat.getIntegerInstance().format(sum);
+  }
```
See above.

wypoon commented:
I have been on a break. Just returning to this. I'm rebasing on master since there are conflicts. I'll follow up after that.

wypoon commented:
I ran `IcebergSourceParquetPosDeleteBenchmark` and `IcebergSourceParquetEqDeleteBenchmark` on the state of master when I last rebased and on my last change. The results for `readIceberg` and `readIcebergVectorized` are as follows:
<img width="1395" alt="Screen Shot 2022-08-03 at 11 13 05 AM" src="https://user-images.githubusercontent.com/3925490/182683434-1f1a6583-703f-4b96-a671-9049a1ea46ab.png">

There is hardly any difference for `IcebergSourceParquetPosDeleteBenchmark`.
For `IcebergSourceParquetEqDeleteBenchmark`, the numbers show greater difference (some worse and some better, comparing after to before), but due to the much greater variances, I think the differences are within the variances.

wypoon commented:
I'm going to rebase on master again and conform to the new spotless code style so that this can merge.

wypoon commented:
@RussellSpitzer @flyrain I have rebased on master. @flyrain I have fixed the bug you pointed out; thank you for pointing it out.
Please review this again. I kept the changes in Spark 3.2, since I was rebasing. I can port the changes to Spark 3.3 in this same PR (and am inclined to do so), if the changes look good to you.

wypoon commented:
I have ported the changes to Spark 3.3. The Spark 3.3 changes are identical to the Spark 3.2 changes (the affected files are identical at this point).

flyrain commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchReader.java:
```diff
@@ -170,6 +173,7 @@ Pair<int[], Integer> posDelRowIdMapping() {
      * @return the mapping array and the new num of rows in a batch, null if no row is deleted
      */
     Pair<int[], Integer> buildPosDelRowIdMapping(PositionDeleteIndex deletedRowPositions) {
+      LOG.debug("Building row id mapping from positional deletes");
```
Nit: is the log necessary? There would be a lot of messages in a real use case when debug level is on. I doubt how useful it is.

wypoon replied:
I'm fine with removing the debug logging for the method calls in `ColumnBatchReader`. I'll do so.

wypoon replied:
Removed.

flyrain commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchReader.java:
```diff
@@ -183,6 +187,8 @@ Pair<int[], Integer> buildPosDelRowIdMapping(PositionDeleteIndex deletedRowPosit
           currentRowId++;
         } else if (hasIsDeletedColumn) {
           isDeleted[originalRowId] = true;
+        } else {
+          deletes.incrementDeleteCount();
         }
```
We still need the metrics in case of `_deleted` column, right?
```
else {
     if (hasIsDeletedColumn) {
          isDeleted[originalRowId] = true;
     }
    deletes.incrementDeleteCount();
}
```

wypoon replied:
So, following the explanation by @RussellSpitzer: 

> there is just now a second read path with _is_deleted metadata column which actually returns deleted rows. When we follow that path I'm not sure it's important for us to count the deleted rows since we'll be returning them anyway, I could go either way. 

I decided that for this read path with the _id_deleted metadata column, we will not perform the count. That is what has been implemented here.

Pardon my ignorance: for my benefit, how does one perform such a read (where `hasIsDeletedColumn` is true)?

flyrain replied:
It is triggered by either projection or filtering. Here are two examples:
```
select name, _deleted from student
```
```
select name from student where _deleted=true
```
It is true that user can get row-level deleted rows, users can also get both deleted and un-deleted rows with `_deleted` column, the first sql above returns that. The way I saw it is that the delete count metric is orthogonal to the metadata column `_deleted`. It reports how many row-level deletes applied as its label indicates, `public static final String DISPLAY_STRING = "number of row deletes applied"`. I don’t have strong option on it though. What do you think? BTW, the non-vectorized read also didn’t report the metric in case of `_deleted` column, see line 255 in class `DeleteFilter`.

wypoon replied:
@flyrain thank you for explaining how reads with `hasIsDeletedColumn` occur. 
I didn't really understand that before, but based on @RussellSpitzer's comment I cited, I figured it was fine not to count deletes for such reads. I have done this consistently for both vectorized and non-vectorized reads, as you realize. This is deliberate.

Now that I understand the use case for `hasIsDeletedColumn`, I agree that there is merit to the argument for considering "number of row deletes applied" metric to apply regardless of `hasIsDeletedColumn`. On the other hand, for the queries where `hasIsDeletedColumn` is true, you are able to infer the number of row deletes applied from the result. I don't have a strong opinion on what to do in the `hasIsDeletedColumn` scenario. But the current implementation is consistent. @RussellSpitzer do you want to weigh in? 

If we want to count the deletes always, we'll need to modify `DeleteFilter#markDeleted` (we'll need to pass in the `DeleteCounter`) as well as the changes you mention for `ColumnarBatchReader`.

wypoon replied:
I was mistaken about needing to modify `Deletes#markDeleted`, `Deletes#streamingMarker` and `Deletes.PositionStreamDeleteMarker`. I just needed to change `BaseReader.SparkDeleteFilter#markRowDeleted` to increment the counter there. `BaseReader.SparkDeleteFilter` keeps a `DeleteCounter` instance.

flyrain commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchReader.java:
```diff
@@ -229,6 +236,8 @@ void applyEqDelete(ColumnarBatch columnarBatch) {
           currentRowId++;
         } else if (hasIsDeletedColumn) {
           isDeleted[rowIdMapping[rowId]] = true;
+        } else {
+          deletes.incrementDeleteCount();
```
The same comment here.

flyrain commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchReader.java:
```diff
@@ -217,6 +223,7 @@ int[] initEqDeleteRowIdMapping() {
      * @param columnarBatch the {@link ColumnarBatch} to apply the equality delete
      */
     void applyEqDelete(ColumnarBatch columnarBatch) {
+      LOG.debug("Applying equality deletes to row id mapping");
```
The same comment here.

wypoon replied:
Removed.

flyrain commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseReader.java:
```diff
@@ -77,6 +78,7 @@
   private final NameMapping nameMapping;
   private final ScanTaskGroup<TaskT> taskGroup;
   private final Iterator<TaskT> tasks;
+  private final DeleteCounter counter;
```
+1 to put it here

flyrain commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:
```diff
@@ -130,7 +140,8 @@ protected Table createTable(String name, Schema schema, PartitionSpec spec) {
     ops.commit(meta, meta.upgradeToFormatVersion(2));
-    if (vectorized) {
+    table.updateProperties().set(TableProperties.DEFAULT_FILE_FORMAT, format).commit();
+    if (format.equals("parquet") && vectorized) {
```
This is all about parquet format. We may do this 
```
 if (format.equals("parquet")) {
   if (vectorized) {
      table
          .updateProperties()
          .set(TableProperties.PARQUET_VECTORIZATION_ENABLED, "true")
          .set(
              TableProperties.PARQUET_BATCH_SIZE,
              "4") // split 7 records to two batches to cover more code paths
          .commit();
    } else {
      table.updateProperties().set(TableProperties.PARQUET_VECTORIZATION_ENABLED, "false").commit();
    }
}
```
Also since we added orc and avro, we may set the vectorized property explicitly like we did for `parquet` other than using the default value. The default value may change in the future, which will break the tests.

wypoon replied:
I could have been mistaken, but when I started working on this, vectorized read was only supported for parquet. For this reason, the parameters for the testsuite only has vectorized=true for format=parquet. In any case, the parameters give a fair coverage of both file formats and vectorized/non-vectorized.

I think I understand your concern. If I understand it correctly, your concern is with `TableProperties.ORC_VECTORIZATION_ENABLED` which currently defaults to false. If format=orc, we want to set `TableProperties.ORC_VECTORIZATION_ENABLED` explicitly in case the default changes in the future.

I'll fix this.

wypoon replied:
Fixed.

flyrain reviewed (commented):
Sorry for the delay. Super busy these days. The performance test looks good. I think we are getting close.

wypoon commented:
@flyrain thanks for reviewing this again. I responded to your comments. Aside from the issue regarding counting the deletes for the case where there is a _is_deleted metadata column, I'll go ahead and make the changes.

wypoon commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseReader.java:
```diff
@@ -261,7 +268,10 @@ protected InputFile getInputFile(String location) {
     @Override
     protected void markRowDeleted(InternalRow row) {
-      row.setBoolean(columnIsDeletedPosition(), true);
+      if (!row.getBoolean(columnIsDeletedPosition())) {
+        row.setBoolean(columnIsDeletedPosition(), true);
+        counter().increment();
+      }
```
`markRowDeleted` can be called as the `Consumer` in `Deletes.markDeleted(CloseableIterable<T>, Predicate<T>, Consumer<T>)` via `DeleteFilter#applyPosDeletes(CloseableIterable<T>)` and via `DeleteFilter#applyEqDeletes(CloseableIterable<T>)`.
I assume we want to avoid counting twice when the same row is deleted via a positional delete and then via an equality delete.
Thus we check to see if the row is already marked deleted, and if it is not, then we mark it and increment the count.
This count is tested in `TestSparkReaderDeletes#testMixedPosAndEqDeletesWithDeletedColumn`.

wypoon commented:
@RussellSpitzer @flyrain I have updated the PR.
I went ahead and implemented the necessary changes so that the number of deletes applied is counted whether or not there is a `_deleted` metadata column in the read.
I have also addressed the other feedback.

flyrain commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:
```diff
@@ -130,16 +140,19 @@ protected Table createTable(String name, Schema schema, PartitionSpec spec) {
+    } else if (format.equals("parquet")) { // in this case, non-vectorized
       table.updateProperties().set(TableProperties.PARQUET_VECTORIZATION_ENABLED, "false").commit();
+    } else if (format.equals("orc")) { // we only have non-vectorized for orc in our parameters
+      table.updateProperties().set(TableProperties.ORC_VECTORIZATION_ENABLED, "false").commit();
```
Minor suggestion: this will cover all cases, and no comment is needed.
```
    if (format.equals("parquet")) {
      table.updateProperties().set(TableProperties.PARQUET_VECTORIZATION_ENABLED, String.valueOf(vectorized)).commit();
      if (vectorized) {
        table.updateProperties().set(TableProperties.PARQUET_BATCH_SIZE, "4").commit();
      }
    } else if (format.equals("orc")) {
      table.updateProperties().set(TableProperties.ORC_VECTORIZATION_ENABLED, String.valueOf(vectorized)).commit();
      if (vectorized) {
        table.updateProperties().set(TableProperties.ORC_BATCH_SIZE, "4").commit();
      }
    }
```

wypoon replied:
Thanks for the suggestion. I have adopted a variant of it.

flyrain commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:
```diff
@@ -362,6 +388,8 @@ public void testEqualityDeleteWithDeletedColumn() throws IOException {
         rowSet(tableName, PROJECTION_SCHEMA.asStruct(), "id", "data", "_deleted");
 
     Assert.assertEquals("Table should contain expected row", expected, actual);
+    long expectedDeletes = 3L;
+    checkDeleteCount(expectedDeletes);
```
Nit: seems not necessary to have them in two lines, but just my personal preference.

wypoon replied:
Fixed.

flyrain reviewed (commented):
LGTM overall. Thanks @wypoon.

wypoon commented:
@flyrain thank you for reviewing. I hope this can be merged now.

flyrain commented:
Great. Thanks @wypoon for working on this.  Hi @RussellSpitzer and @kbendick, do you want to take another look and approve it? There is an authentication issue from my side that I still don't have the permission to merge even though I'm a committer now. Hopefully, my authentication issue could be resolved next week.

kbendick commented:
> Great. Thanks @wypoon for working on this. Hi @RussellSpitzer and @kbendick, do you want to take another look and approve it? There is an authentication issue from my side that I still don't have the permission to merge even though I'm a committer now. Hopefully, my authentication issue could be resolved next week.

Yeah sure let me do another pass. Congrats again on becoming a committer!

flyrain commented:
Thanks @kbendick!

kbendick commented on core/src/main/java/org/apache/iceberg/deletes/DeleteCounter.java:
```diff
@@ -0,0 +1,35 @@
+  /** Return the current value of the counter. */
+  public long get() {
+    return count;
+  }
```
I can’t recall if this was covered or not already, but is there a reason we can’t use `Counter` from the `metrics` package in api?

I think this work might have been done before the `Counter` class from api existed, but there’s also very likely a reason to stick with this simpler class (for Spark serde or something).

https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/metrics/Counter.java

kbendick replied:
I’m not sure how the entire `MetricsContext` integrates with Spark and executors etc.

If we don’t have existing work that uses `MetricsContext` for custom Spark metrics (particularly task level metrics), possibly we can update this code in a later PR.

I know Steven handled a lot of the metrics context adjustments for Flink and @nastra did a lot of the work for that in general.

wypoon replied:
This wasn't raised previously. I wasn't aware of `org.apache.iceberg.metrics.Counter` and in fact, it didn't exist when I started this work several months ago (although `org.apache.iceberg.metrics.MetricsContext.Counter` did exist).
I think `org.apache.iceberg.metrics.Counter` is a more complex API than needed here (in particular, we do not need a name or a unit), although one could use it. I'd prefer not to refactor the code. A more generic counter could be used, but `DeleteCounter` is specific to this use case and is for internal bookkeeping only.

kbendick replied:
Agreed. And I can see how the MetricsContext work took place alongside this work (or realistically after it).

Unifying the two should be kept in mind, but the `DeleteCounter` is a well named, specific class that can’t really be mistaken for something else and can be updated later.

Thanks for clarifying @wypoon

wypoon replied:
Yes, I think exposing this through `MetricsContext` could be valuable and such work would be a separate PR; at that time, the `DeleteCounter` could be replaced by the more general `Counter`.

kbendick commented on data/src/test/java/org/apache/iceberg/data/DeleteReadTests.java:
```diff
@@ -282,6 +306,7 @@ public void testEqualityDeletesSpanningMultipleDataFiles() throws IOException {
         FileHelpers.writeDataFile(table, Files.localOutput(temp.newFile()), Row.of(0), records);
 
     table.newAppend().appendFile(dataFile).commit();
+    // At this point, the table has 7 + 8 = 15 rows, of which all but one are in duplicate.
```
Nit / non-blocking: it’s a little strange to me to have the whole line comments with empty lines afterwards.

Since this is just a test and it said “at this point”, I feel like we could move it to be directly above the next code block. But it’s just test code so I absolutely wouldn’t block on this.

wypoon replied:
Perhaps I'm idiosyncratic. I put the comment after the statement because the "at this point" refers to the state after the statement has executed and not before. (If the comment were short enough, I would have put it on the same line as the statement, after it. That would probably be more natural.)

Nevertheless, I have now put the comments before the statements.

kbendick commented on data/src/test/java/org/apache/iceberg/data/DeleteReadTests.java:
```diff
@@ -297,11 +322,14 @@ public void testEqualityDeletesSpanningMultipleDataFiles() throws IOException {
             table, Files.localOutput(temp.newFile()), Row.of(0), dataDeletes, deleteRowSchema);
 
     table.newRowDelta().addDeletes(eqDeletes).commit();
+    // At ths point, the table has (7 - 3) + (8 - 4) = 8 rows. 7 rows in all are deleted.
```
Nit: the phrasing of this comment is a little hard for me to follow without reading it twice.

Maybe something like `// At ths point, the table has two data files. They have 7 rows and 8 rows with 3 deleted rows and 4 deleted rows respectively.`

But again this is just a test so I wouldn’t block on this.

kbendick commented on data/src/test/java/org/apache/iceberg/data/DeleteReadTests.java:
```diff
@@ -297,11 +322,14 @@ public void testEqualityDeletesSpanningMultipleDataFiles() throws IOException {
     StructLikeSet actual = rowSet(tableName, table, "*");
 
     Assert.assertEquals("Table should contain expected rows", expected, actual);
+    // 3 deletes in the first data file and 4 deletes in the second data file
```
Nit / Non-blocking: this comment seems redundant given we’ve mentioned the 7 deletes previously. But it’s not incorrect so I’m not opposed to keeping it.

wypoon replied:
Removed this comment as above comment has made it clear how many deletes there are.

kbendick commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchReader.java:
```diff
@@ -181,8 +181,11 @@ Pair<int[], Integer> buildPosDelRowIdMapping(PositionDeleteIndex deletedRowPosit
+          if (hasIsDeletedColumn) {
+            isDeleted[originalRowId] = true;
+          }
+          deletes.incrementDeleteCount();
```
Are we sure that this `else` block is refactored correctly?

Is it possible for `hasIsDeletedColumn` to be false but we still need to increment `deletes.incrementDeleteCount()` (at least in the context of position deletes)?

flyrain replied:
Here is our discussion on this, https://github.com/apache/iceberg/pull/4588#discussion_r945959579. In general, I don't have strong option on either side, but prefer think this metric is orthogonal to metadata column `_deleted`.

kbendick replied:
Oh I see. Yeah then we’d want to keep the metric whether we had the column or not.

Thanks for pointing me in the right direction.

wypoon replied:
Yes, this is correct.
The structure of the logic is
```
        if (...) {
          // row is not deleted
        } else {
          // row is deleted
          if (hasIsDeletedColumn) { ... }
          deletes.incrementDeleteCount();
        }
```
In the else clause, we increment the counter regardless of `hasIsDeletedColumn` because this is the case where the row is deleted.

kbendick replied:
One small nit: Can we add a blank line between the if block and the metrics increment? It’s in-line with the style guide and I think would reduce the tendency to wonder if the final else is intentional.

wypoon replied:
Sure. I have added blank lines after non-final if blocks.

kbendick commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/BatchDataReader.java:
```diff
@@ -49,6 +53,7 @@ protected Stream<ContentFile<?>> referencedFiles(FileScanTask task) {
   @Override
   protected CloseableIterator<ColumnarBatch> open(FileScanTask task) {
     String filePath = task.file().path().toString();
+    LOG.debug("Opening data file {}", filePath);
```
Nit: This debug log could arguably be removed, but I wouldn’t block on this comment. It could be good to have more verbose debug logs.

kbendick reviewed (commented):
Overall this looks good to me, though I’m curious if it’s possible to use `MetricsContext` for this.

I don’t personally want to block on that as this has been outstanding for some time, but figured it’s worth bringing up.

Otherwise, this is a very welcomed addition and I’m happy to have the metrics displayed in the UI.

kbendick reviewed (approved):
Looking at it further, it looks like integrating MetricsContext into Spark (and Spark’s custom metrics) would possibly take a bit more work.

I don’t see any reason to hold this work up any longer while that is done.

This added metric has very real user-facing value now, and unifying the two metrics can be done in a follow up PR if possible.

The `DeleteCounter` is appropriately named and has a relatively small footprint and changing out that for `MetricsContext` can be done at a later date. I think it’s definitely time to get this work merged, as @wypoon has been carrying it for quite some time.

And again, this patch has so much user facing value that I don’t see a need to hold it up any further as it doesn’t seem to complicate anything in a way that can’t be changed later and is indeed pretty light-weight bookkeeping.

Thanks @wypoon for your diligent work on this important feature!

wypoon commented:
@kbendick thanks for reviewing. I have addressed some minor nits.

wypoon commented:
@flyrain @kbendick since you have approved this, can one of you please merge it? (The only update since is one adding some blank lines after if blocks and tweaking of comments in the test class.)

flyrain commented:
@wypoon, sorry for the delay, working on my apache authentication issue. I don't have the permission to merge until the issue is resolved. May @RussellSpitzer can help?

kbendick commented:
> @flyrain @kbendick since you have approved this, can one of you please merge it? (The only update since is one adding some blank lines after if blocks and tweaking of comments in the test class.)

Unfortunately I’m not able to merge yet as I’m not a committer. Hopefully either @flyrain’s credentials will get worked out or somebody else will handle it!

And sorry for any delay in response, I’ve been out very sick this week but trying to stay on top of things when I have the energy. 😷

pvary commented on core/src/main/java/org/apache/iceberg/deletes/Deletes.java:
```diff
@@ -62,6 +62,7 @@ public static <T> CloseableIterable<T> filter(
     return equalityFilter.filter(rows);
   }
 
+  /** Returns the same rows that are input, while marking the deleted ones. */
```
Maybe this is only me, but when we add a comment,  we can create a full javadoc. I personally hate partial Java doc comments.

wypoon replied:
Fixed.

pvary commented on core/src/main/java/org/apache/iceberg/deletes/Deletes.java:
```diff
@@ -74,9 +75,24 @@ public static <T> CloseableIterable<T> markDeleted(
   }
 
+  /**
+   * Returns the remaining rows (the ones that are not deleted), while counting the deleted ones.
```
Please add the description for the parameters

wypoon replied:
Fixed.

pvary commented on core/src/main/java/org/apache/iceberg/deletes/Deletes.java:
```diff
@@ -74,9 +75,24 @@ public static <T> CloseableIterable<T> markDeleted(
+            boolean deleted = isDeleted.test(item);
+            if (deleted) {
+              counter.increment();
+            }
```
Nit: new line after the block

pvary commented on core/src/main/java/org/apache/iceberg/deletes/Deletes.java:
```diff
@@ -215,19 +239,27 @@ boolean isDeleted(T row) {
+          boolean deleted = isDeleted(item);
+          if (deleted) {
+            counter.increment();
+          }
```
Nit: new line

pvary commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/metrics/NumDeletes.java:
```diff
@@ -0,0 +1,46 @@
+    long sum = initialValue;
+    for (int i = 0; i < taskMetrics.length; i++) {
+      sum += taskMetrics[i];
+    }
```
Nit: newline

pvary commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/metrics/NumSplits.java:
```diff
@@ -0,0 +1,44 @@
+    long sum = initialValue;
+    for (int i = 0; i < taskMetrics.length; i++) {
+      sum += taskMetrics[i];
+    }
```
Nit: newline

pvary commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/SparkSQLExecutionHelper.java:
```diff
@@ -0,0 +1,68 @@
+        attempts -= 1;
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
```
Nit: newline

pvary commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/SparkSQLExecutionHelper.java:
```diff
@@ -0,0 +1,68 @@
+    while (lastExecution.metricValues() == null && attempts > 0) {
+      try {
+        Thread.sleep(100);
+        attempts -= 1;
```
Nit: Why not `attempts--`  or `--attempts`?

wypoon replied:
A scala-ism. :-)

pvary commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/metrics/NumSplits.java:
```diff
@@ -0,0 +1,44 @@
+    long sum = initialValue;
+    for (int i = 0; i < taskMetrics.length; i++) {
+      sum += taskMetrics[i];
+    }
```
Nit: newline

pvary commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/SparkSQLExecutionHelper.java:
```diff
@@ -0,0 +1,68 @@
+        attempts -= 1;
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
```
Nit: newline

pvary commented:
@wypoon: LGTM. Left a few formatting asks, nothing big. For the next time, AFAIK we do separate PRs for the different versions. We implement stuff first for the latest version and then backport to the older ones. This is generally easier for the reviewer.
Thanks for the PR.
Peter

wypoon commented:
Ok, local build succeeds, but I guess the CI must rebase on master; even though there are no conflicts, the build fails, because of a new class added by #5578 and #5682.

wypoon commented:
Rebased on master, and fixed the newly added `ChangelogRowReader`'s use of `SparkDeleteFilter`. @flyrain can you please take a quick look at that small change? Otherwise, the only other changes are in formatting and comments.

flyrain commented:
LGTM. Thanks @wypoon.

wypoon commented:
@flyrain have you sorted out your credentials? can you or another committer merge this now? Thanks.

flyrain commented:
@wypoon I'd love to help on here. Unfortunately the issue is still there :-(. I wish secretary@apache.org could be a bit more responsive.

pvary commented:
Thanks @wypoon for the PR and @flyrain, @kbendick and @RussellSpitzer for the reviews!

wypoon commented:
Thanks @pvary!
